Check for missing containers and rebuild them

schrej opened the pull request:
fixes Pterodactyl/Panel#148

Check for missing containers when the daemon initializes or a server is started and trigger a rebuild if the container is not found.

On daemon initialize the rebuild is done asynchronous.

schrej committed:
only remove old container if it exists on rebuild

schrej committed:
attempt to rebuild container if it is not found on server creation

schrej committed:
when starting mark server for rebuild if container is not found and start again
add necessary error messages to putServerPower route

DaneEveritt commented on src/controllers/docker.js:
```diff
@@ -393,11 +393,23 @@ class Docker {
             this.server.log.debug('Removing old server container...');
             // @TODO: logic here to determine if this failed, if so we need to
```
Not really a to-do anymore. :)

schrej replied:
Well it seemed to describe something else so i kept it. If you meant the check i implemented i can remove it.

DaneEveritt commented on src/controllers/server.js:
```diff
@@ -72,13 +72,17 @@ class Server extends EventEmitter {
+                this.rebuild(() => {
+                    this.log.fatal('Could not recreate container.');
+                });
+            } else return next(err);
```
Was this eslinted? I don't believe this syntax — `} else return next(err);` — is valid per our style preferences.

schrej replied:
It was eslinted (i think thats the linter i installed for atom, it was really agressive about everything :D) but it didn't complain. How do you like it? just a linebreak or with brackets?

DaneEveritt replied:
Interesting, I never considered doing it that way. I'd prefer a line break with brackets simply to match the rest of the code.

DaneEveritt commented on src/controllers/server.js:
```diff
@@ -215,6 +219,16 @@ class Server extends EventEmitter {
+                        if (modifyError) return this.log.error('Could not modify config.');
+                        this.start(_.noop); // Ignore the callback as there is nowhere to send the errors to.
+                    });
+                    return next(new Error('Server container was not found and needs to be rebuilt. Your request has been accepted and will be processed once the rebuild is complete.'));
```
Is the request actually going to be processed here, or simply rebuild the server and then do nothing? In any case, I think we should probably return without an error, since the rebuild will still continue to a boot.

schrej replied:
It just changes the configuration to queue the server for container rebuild and then starts the server again. The container is rebuilt and the server starts.
I'm throwing the error as [you are doing the same](https://github.com/Pterodactyl/Daemon/blob/develop/src/controllers/server.js#L195) when a request is sent and the server is queued for container rebuild. The error is returned to the putServerPower route. This is why i drop the new callback from this.start() as well, there is no open http connection where i could return it to.

DaneEveritt commented:
Currently testing this PR.

DaneEveritt commented on src/controllers/server.js:
```diff
@@ -72,13 +72,17 @@ class Server extends EventEmitter {
+                this.log.info('Container was not found. Attempting to recreate it.');
+                next(); // Continue normal initialization
+                this.rebuild(() => {
+                    this.log.fatal('Could not recreate container.');
```
This logic fails, and will trigger a fatal error even if the rebuild was successful. You should be looking for an `err` object here, and throwing that error into the `this.log.fatal()` function if it exists, otherwise do nothing.

As it stands, even a successful rebuild will throw that error to the console.

```
16:27:03.175Z DEBUG wings: New container created successfully, updating config... (server=67c7a4b0-6f62-46c4-af32-6494d7871ccf)
16:27:03.189Z DEBUG wings: Loaded configuration and initalized server. (server=67c7a4b0-6f62-46c4-af32-6494d7871ccf)
16:27:03.190Z FATAL wings: Could not recreate container. (server=67c7a4b0-6f62-46c4-af32-6494d7871ccf)
```

schrej replied:
Oh yeah sorry, this was the last thing i did yesterday as i was waiting for your response.

DaneEveritt commented on src/controllers/server.js:
```diff
@@ -215,6 +219,16 @@ class Server extends EventEmitter {
             if (err) {
+                if (err.statusCode === 404) { // container not found
+                    this.log.error('The container for this server could not be found. Trying to rebuild it.');
+                    this.emit('console', `${Ansi.style.red}(Daemon) The container of this server could not be found and has to be rebuilt.`);
```
We can probably skip this console report, the rebuild will output information to the screen anyways, and this will probably scare users when in reality it will most likely not be a huge impact.

DaneEveritt commented on src/controllers/docker.js:
```diff
@@ -393,11 +393,23 @@ class Docker {
+                // if it doesn't we'll just skip removal
+                } else {
+                    this.server.log.debug('Old container not found, skipping.');
+                    next(false, newContainerInfo);
```
Should return `null` not a boolean.

DaneEveritt commented:
Other than the commented things, this PR is fantastic and works well. Will pull it once those few things are fixed. 👍

schrej committed:
remove resolved todo

pass null instead of false for unnecessary callback parameter

schrej committed:
only show fatal error when rebuild fails on server init

change container not found console output to cyan as it is not really an error
adapt proposed code style

DaneEveritt commented on src/controllers/server.js:
```diff
@@ -221,7 +223,7 @@ class Server extends EventEmitter {
                 if (err.statusCode === 404) { // container not found
                     this.log.error('The container for this server could not be found. Trying to rebuild it.');
-                    this.emit('console', `${Ansi.style.red}(Daemon) The container of this server could not be found and has to be rebuilt.`);
+                    this.emit('console', `${Ansi.style.cyan}(Daemon) The container of this server could not be found and needs to be rebuilt.`);
```
I'd say just completely remove this since the next line is one saying the container is being rebuilt.

schrej replied:
i would keep it so people won't get confused by untriggered rebuilds but ok

DaneEveritt commented on src/controllers/server.js:
```diff
@@ -79,10 +79,12 @@ class Server extends EventEmitter {
                 next(); // Continue normal initialization
-                this.rebuild(() => {
-                    this.log.fatal('Could not recreate container.');
+                this.rebuild((rebuildErr) => {
```
This fails eslint checking, `rebuildErr` should not be surrounded by parentheses.

schrej replied:
yeah, thats why you don't use the github online editor 😬

schrej committed:
remove special container not found message on start
change container rebuild message to be more generic
fix linting issues
